// @augment-vir/common/dist/cjs/augments/string/suffixes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeSuffix = exports.addSuffix = exports.removePercent = exports.addPercent = exports.removePx = exports.addPx = exports.pxSuffix = exports.percentSuffix = void 0;
const common_number_1 = require("../common-number");
exports.percentSuffix = '%';
exports.pxSuffix = 'px';
function addPx(input) {
    return addSuffix({ value: input, suffix: exports.pxSuffix });
}
exports.addPx = addPx;
function removePx(input) {
    return (0, common_number_1.toEnsuredNumber)(removeSuffix({ value: input, suffix: exports.pxSuffix }));
}
exports.removePx = removePx;
function addPercent(input) {
    return addSuffix({ value: input, suffix: exports.percentSuffix });
}
exports.addPercent = addPercent;
function removePercent(input) {
    return (0, common_number_1.toEnsuredNumber)(removeSuffix({ value: input, suffix: exports.percentSuffix }));
}
exports.removePercent = removePercent;
function addSuffix({ value, suffix, }) {
    if (String(value).endsWith(suffix)) {
        return String(value);
    }
    else {
        return `${String(value)}${suffix}`;
    }
}
exports.addSuffix = addSuffix;
function removeSuffix({ value, suffix, }) {
    if (value.endsWith(suffix)) {
        return value.substring(0, value.length - suffix.length);
    }
    else {
        return value;
    }
}
exports.removeSuffix = removeSuffix;
