// @augment-vir/common/dist/cjs/augments/string/suffixes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pxSuffix = exports.percentSuffix = void 0;
exports.addPx = addPx;
exports.removePx = removePx;
exports.addPercent = addPercent;
exports.removePercent = removePercent;
exports.addSuffix = addSuffix;
exports.removeSuffix = removeSuffix;
const common_number_1 = require("../common-number");
exports.percentSuffix = '%';
exports.pxSuffix = 'px';
function addPx(input) {
    return addSuffix({ value: input, suffix: exports.pxSuffix });
}
function removePx(input) {
    return (0, common_number_1.toEnsuredNumber)(removeSuffix({ value: input, suffix: exports.pxSuffix }));
}
function addPercent(input) {
    return addSuffix({ value: input, suffix: exports.percentSuffix });
}
function removePercent(input) {
    return (0, common_number_1.toEnsuredNumber)(removeSuffix({ value: input, suffix: exports.percentSuffix }));
}
function addSuffix({ value, suffix, }) {
    if (String(value).endsWith(suffix)) {
        return String(value);
    }
    else {
        return `${String(value)}${suffix}`;
    }
}
function removeSuffix({ value, suffix, }) {
    if (value.endsWith(suffix)) {
        return value.substring(0, value.length - suffix.length);
    }
    else {
        return value;
    }
}
