// @augment-vir/common/dist/cjs/augments/object/object-entries.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isKeyof = isKeyof;
exports.getObjectTypedKeys = getObjectTypedKeys;
exports.getObjectTypedValues = getObjectTypedValues;
exports.getObjectTypedEntries = getObjectTypedEntries;
exports.getEntriesSortedByKey = getEntriesSortedByKey;
exports.typedObjectFromEntries = typedObjectFromEntries;
const typed_has_property_1 = require("./typed-has-property");
/** @deprecated This is the same as hasKey */
function isKeyof(key, object) {
    return (0, typed_has_property_1.typedHasProperty)(object, key);
}
function getObjectTypedKeys(input) {
    let reflectKeys;
    try {
        reflectKeys = Reflect.ownKeys(input);
    }
    catch (error) { }
    return (reflectKeys ??
        [
            ...Object.keys(input),
            ...Object.getOwnPropertySymbols(input),
        ]);
}
function getObjectTypedValues(input) {
    return getObjectTypedKeys(input).map((key) => input[key]);
}
function getObjectTypedEntries(input) {
    return getObjectTypedKeys(input).map((key) => [
        key,
        input[key],
    ]);
}
function getEntriesSortedByKey(input) {
    return Object.entries(input).sort((tupleA, tupleB) => tupleA[0].localeCompare(tupleB[0]));
}
function typedObjectFromEntries(entries) {
    return Object.fromEntries(entries);
}
