// @augment-vir/common/dist/cjs/augments/json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.areJsonEqual = exports.stringifyJson = exports.parseJson = void 0;
const run_time_assertions_1 = require("run-time-assertions");
const error_1 = require("./error");
const matches_object_shape_1 = require("./object/matches-object-shape");
const object_1 = require("./object/object");
function parseJson({ jsonString, errorHandler, shapeMatcher, }) {
    try {
        const parsedJson = JSON.parse(jsonString);
        if (shapeMatcher != undefined) {
            if ((0, run_time_assertions_1.isRunTimeType)(shapeMatcher, 'object')) {
                (0, matches_object_shape_1.assertMatchesObjectShape)(parsedJson, shapeMatcher);
            }
            else {
                (0, run_time_assertions_1.assertRunTimeType)(parsedJson, (0, run_time_assertions_1.getRunTimeType)(shapeMatcher), 'parsedJson');
            }
        }
        return parsedJson;
    }
    catch (error) {
        if (errorHandler) {
            return errorHandler(error);
        }
        else {
            throw error;
        }
    }
}
exports.parseJson = parseJson;
function stringifyJson({ source, whitespace, errorHandler, }) {
    try {
        const stringifiedJson = JSON.stringify(source, undefined, whitespace);
        return stringifiedJson;
    }
    catch (error) {
        if (errorHandler) {
            return errorHandler(error);
        }
        else {
            throw error;
        }
    }
}
exports.stringifyJson = stringifyJson;
const areJsonEqualFailureMessage = 'Failed to compare objects using JSON.stringify';
function baseAreJsonEqual(a, b, ignoreStringifyErrors) {
    return (stringifyJson({
        source: a,
        errorHandler(error) {
            if (ignoreStringifyErrors) {
                return '';
            }
            else {
                throw error;
            }
        },
    }) ===
        stringifyJson({
            source: b,
            errorHandler(error) {
                if (ignoreStringifyErrors) {
                    return '';
                }
                else {
                    throw error;
                }
            },
        }));
}
function areJsonEqual(a, b, options = {}) {
    try {
        if (a === b) {
            return true;
        }
        if ((0, object_1.isObject)(a) && (0, object_1.isObject)(b)) {
            const areKeysEqual = baseAreJsonEqual(Object.keys(a).sort(), Object.keys(b).sort(), !!options?.ignoreNonSerializableProperties);
            if (!areKeysEqual) {
                return false;
            }
            return Object.keys(a).every((keyName) => {
                return areJsonEqual(a[keyName], b[keyName]);
            });
        }
        else {
            return baseAreJsonEqual(a, b, !!options?.ignoreNonSerializableProperties);
        }
    }
    catch (caught) {
        const error = (0, error_1.ensureError)(caught);
        if (error.message.startsWith(areJsonEqualFailureMessage)) {
            throw error;
        }
        error.message = `${areJsonEqualFailureMessage}: ${error.message}`;
        throw error;
    }
}
exports.areJsonEqual = areJsonEqual;
