// @augment-vir/common/dist/cjs/augments/array/array.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.filterMap = exports.arrayToObject = exports.groupArrayBy = exports.repeatArray = exports.typedMap = exports.typedArrayIncludes = exports.trimArrayStrings = exports.flatten2dArray = exports.filterOutIndexes = void 0;
const get_or_set_1 = require("../object/get-or-set");
const object_entries_1 = require("../object/object-entries");
function filterOutIndexes(array, indexes) {
    return array.filter((_, index) => !indexes.includes(index));
}
exports.filterOutIndexes = filterOutIndexes;
function flatten2dArray(array2d) {
    const flattened = array2d.reduce((accum, row) => accum.concat(row), []);
    return flattened;
}
exports.flatten2dArray = flatten2dArray;
function trimArrayStrings(input) {
    return input.map((line) => line.trim()).filter((line) => line !== '');
}
exports.trimArrayStrings = trimArrayStrings;
function typedArrayIncludes(array, input) {
    return array.includes(input);
}
exports.typedArrayIncludes = typedArrayIncludes;
/** Preserves tuple types. */
function typedMap(arrayToMap, mapCallback) {
    return arrayToMap.map(mapCallback);
}
exports.typedMap = typedMap;
function repeatArray(repeatCount, array) {
    return Array.from({ length: repeatCount }, () => [...array]).flat();
}
exports.repeatArray = repeatArray;
/**
 * Polyfill for `Object.groupBy`:
 * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/groupBy
 */
function groupArrayBy(inputArray, callback) {
    return inputArray.reduce((accum, entry, index, originalArray) => {
        const key = callback(entry, index, originalArray);
        const entryArray = (0, get_or_set_1.getOrSet)(accum, key, () => []);
        entryArray.push(entry);
        return accum;
    }, {});
}
exports.groupArrayBy = groupArrayBy;
/**
 * Like `groupArrayBy` but maps array entries to a single key. Meaning, the resulting object does
 * not have an array of elements (unless the original array itself contains arrays).
 */
function arrayToObject(inputArray, callback) {
    return (0, object_entries_1.typedObjectFromEntries)(inputArray.map((entry, index, originalArray) => {
        const key = callback(entry, index, originalArray);
        return [
            key,
            entry,
        ];
    }));
}
exports.arrayToObject = arrayToObject;
function filterMap(inputArray, mapCallback, filterCallback) {
    return inputArray.reduce((accum, entry, index, originalArray) => {
        const mapOutput = mapCallback(entry, index, originalArray);
        const filterOutput = filterCallback(mapOutput, entry, index, originalArray);
        if (filterOutput) {
            accum.push(mapOutput);
        }
        return accum;
    }, []);
}
exports.filterMap = filterMap;
